### ISSUE TITLE
First chance odata exception resolving alternate keys

### ISSUE THREAD

**Issue by mhotze** [OData/odata.net#2221]
When a model contains alternate keys, you have to deal with a first chance odata exception for resolving each. Consider a model with many hundreds of entities each having at least one alternate key. Then it is annoying to debug with catching first chance exceptions on.

### Assemblies affected

OData .Net lib 8.x* (Microsoft.OData.Core)

### Reproduce steps

Debug the test `AlternateKeyShouldWork` with 'break when \<All Common Language Exceptions not in this list\> are thrown' exception setting is activated in Visual Studio. https://github.com/OData/odata.net/blob/64bed784177b9202b73c4510d2461bebf5ad03b7/test/FunctionalTests/Microsoft.OData.Core.Tests/UriParser/ODataUriParserTests.cs#L383

### Expected result

No first chance exception by checking chance of success upfront.

### Actual result

A first chance `ODataException` is caught for each alternate key in the model.

### Additional detail

It is inappropriate that `UriParser/Resolver/AlternateKeysODataUriResolver.ResolveKeys` relies so heavily on trying to resolve the primary key first and then trying to resolve the alternate key by catching prior exception. It would be more smooth to eliminate certain failures by checking the number and name of keys first.

```diff
public override IEnumerable<KeyValuePair<string, object>> ResolveKeys(IEdmEntityType type, IDictionary<string, string> namedValues, Func<IEdmTypeReference, string, object> convertFunc)
{
    IEnumerable<KeyValuePair<string, object>> convertedPairs;
-   try
+   string[] keyNames = type.DeclaredKey.Select(keyProperty => keyProperty.Name).ToArray();
+   if (namedValues.Keys.Count == keyNames.Length)
    {
+       if (namedValues.Keys.All(keyProperty => keyNames.Contains(keyProperty, StringComparer.OrdinalIgnoreCase)))
+       {
            convertedPairs = base.ResolveKeys(type, namedValues, convertFunc);
+       }
    }
-   catch (ODataException)
+   else
    {
        if (!TryResolveAlternateKeys(type, namedValues, convertFunc, out convertedPairs))
        {
-           throw;
+           throw new ODataException(...);
        }
 -  }

    return convertedPairs;
}

```

**Comment by habbes** [comment]
Hi @mhotze I'm looking into this and I think there might be some issues with your proposed changes. First of all, it is checking the names and counts against the declared key, which is part what the `base.ResolveKeys` method does, so it may be a bit redundant. But that may not be a big deal if on average the number of key properties is small.

The other issue is the `base.ResolveKeys` may technically still throw exceptions. And since we're no longer catching them, the exceptions will cause the method to fail even if the alternate key resolver should have been able to successfully resolve the keys. 

I think  safer approach would require duplicating the logic inside `base.ResolveKeys()`, but instead of throwing exceptions, it returns false. Then the alternate key resolution logic can take over. But is that worthwhile?

Out of curiosity, is possible to configure the debugger such that it does not break on this particular exception, or on first-chance exceptions?

**Comment by mhotze** [comment]
Hi @habbes 

checking the count and names seems to be the only way and should definitely be excuted only once.

Regardig the concerns about the other technical exceptions, keep in mind that there won't be any reason to rethrow the once caught exception at all. Just let every exception happen, as long as you can ensure there is no exception caused by the check itself. So all exceptions are actual related only to the primary key.

I agree a pattern like TryResolveKeys with an out parameter would solve the problem, but will change the whole signature what is probably not desired.

Surely one could simply suppress this particular exception, but actual we are interested in catching exactly this particular exception for other purposes out of scope.

**Comment by habbes** [comment]
@mhotze do you depend on the actual exception message, or just that the exception was thrown?

I think another alternative would be to create a `base.TryResolveKeys` method, then refactor the `base.ResolveKeys()` method to depend on the former, and then throw an exception if it returns false. This would avoid duplicating the logic, so it will make it more maintainable, and we can also use the `base.TryResolveKeys` in the `AlternateKeyResolver` and avoid depending on the exception or traversing the collection twice.

The downside to this is that `base.ResolveKeys()` will not know exactly why `TryResolveKeys` failed, so it might just throw the same error message regardless of error. We can use a descriptive error message that accounts for most failure scenarios, but changing the error message is possibly a breaking change in some extreme cases. But if we can overlook that, I think that would be the ideal fix.

**Comment by mhotze** [comment]
@habbes sorry, i should have said, we are interested in breaking on `ODataException` and `ODataErrorException` only in the debugger in order to see odata related problems in general und eventually provide proper `ProblemDetails` in our services.

Finally, i prefer my proposal with a little change the most. Checking count and names in the `ResolveKeys` overload for named values in `AlternateKeysODataUriResolver` before calling `base.ResolveKeys` of `ODataUriResolver`. This duplicate check is more reliable than the one in `base` and should be worthwhile. Nevertheless it's costly if it's still a primary key. with a one third chance it's a primary key, an alternate key or a fault from `base`. As long as the `AlternateKeysODataUriResolver` has to be explicitly injected, the chance for alternate key seams to be quite high. And because the existing `TryResolveAlternateKeys` doesn't throw exceptions, you may swap the order to be sure that exceptions from `base` will be thrown later on.

What about this approach? In `AlternateKeysODataUriResolver` first `TryResolveAlternateKeys` if count and names doesn't match (certainly no primary key), before letting `ODataUriResolver.ResolveKeys` resolve keys and eventually throw expected exceptions.

```diff
public override IEnumerable<KeyValuePair<string, object>> ResolveKeys(IEdmEntityType type, IDictionary<string, string> namedValues, Func<IEdmTypeReference, string, object> convertFunc)
{
    IEnumerable<KeyValuePair<string, object>> convertedPairs;
-   try
+   string[] keyNames = type.DeclaredKey.Select(keyProperty => keyProperty.Name).ToArray();
+   if ((namedValues.Keys.Count != keyNames.Length
+       || namedValues.Keys.Any(keyProperty => !keyNames.Contains(keyProperty, StringComparer.OrdinalIgnoreCase)))
+       && TryResolveAlternateKeys(type, namedValues, convertFunc, out convertedPairs))
+   {
+       return convertedPairs;
+   }
-   catch (ODataException)
+   else
    {
        convertedPairs = base.ResolveKeys(type, namedValues, convertFunc);
-       if (!TryResolveAlternateKeys(type, namedValues, convertFunc, out convertedPairs))
-       {
-          throw;
-       }
    }

    return convertedPairs;
}
```
